### DISCUSSION TITLE
:remove-attr() and :remove-class() pseudo selector operators

### DISCUSSION THREAD

**Discussion by mapx-** [uBlockOrigin/uBlock-issues#2403] [General]
@gorhill 

https://github.com/gorhill/uBlock/commit/992255e9937f81b5dd58524caa2ceeaeb29efb14#commitcomment-92647587

from slack (@MasterKia) 

```
Is this supported?
`##+js(ra, oncontextmenu|onselectstart|ondragstart)`
=>
`##*:remove-attr(oncontextmenu|onselectstart|ondragstart)`
```

We should know / have examples for the new operators

Also uBlock-user question:

```
Is there support for "stay"/"complete" option like we have in the scriptlet version ?

Edit: Will need to use :watch-attr(), seems to be a working replacement for the aformentioned options.
```

**Comment by uBlock-user** [comment]
> ##*:remove-attr(oncontextmenu|onselectstart|ondragstart)

Regex is.

**Comment by gorhill** [comment]
Argument can be text or regex:

    // remove exact oncontextmenu attribute
    ##*:remove-attr(oncontextmenu)

    // remove attributes which match the regexes
    ##*:remove-attr(/oncontextmenu|onselectstart|ondragstart/)
    ##*:remove-attr(/^on[a-z]+/)

And as with other pseudoclass operators, use quotes when parser is tripped.

**Reply by gothic-bum** [reply]
Do you recommend the use of `##*:remove-attr(oncontextmenu)` vs the regex equivalent of `##*:remove-attr(/^oncontextmenu$/)`?

**Reply by gorhill** [reply]
Well for real cases I recommend simplicity and smallest resultset to act on (as with any other operators):

    ...##[oncontextmenu]:remove-attr(oncontextmenu)

**Comment by mapx-** [comment]
@gorhill what about "stay/complete" (added above), should we use `watch-attr` ?

**Reply by gorhill** [reply]
Pseudoclass operators do not need `stay` or `complete` since they are continually enforced, like other operators. But you will need `:watch-attr()` if you want the operators to act when attributes change without layout changes.

**Reply by uBlock-user** [reply]
> Pseudoclass operators do not need stay or complete since they are continually enforced, like other operators.

Without `:watch-attr()`, a filter such as `www.google.com##body[jsaction]:remove-attr(jsaction)` doesn't work, how come ?

**Reply by krystian3w** [reply]
How implemented is `asap` in once or both operators?

https://github.com/gorhill/uBlock/commit/35d7406214e39fa5ad5c73cfab3eecb0eb7c8b7f https://github.com/uBlockOrigin/uAssets/issues/9528

**Reply by krystian3w** [reply]
Also sometimes will be nice use only once without loop/MutationObserver:

https://github.com/FiltersHeroes/PolishAnnoyanceFilters/issues/2667#issuecomment-766357297 (so far I do not have any active example where the loop harms the party).

**Comment by MasterKia** [comment]
https://github.com/uBlockOrigin/uBlock-issues/wiki/Static-filter-syntax#subjectremove-attr-subjectremove-class

**Reply by krystian3w** [reply]
link down?

https://github.com/uBlockOrigin/uBlock-issues/wiki/Static-filter-syntax#subjectremove-attr-subjectremove-class - old

https://github.com/uBlockOrigin/uBlock-issues/wiki/Static-filter-syntax#subjectremove-attrarg-subjectremove-classarg - latest?

**Comment by ghost** [comment]
A case with `:watch-attr()`?

https://www.reddit.com/r/uBlockOrigin/comments/105dkbp/removeclass_works_different_in_comparison_to_rcjs/

**Reply by uBlock-user** [reply]
`class` attribute is very generic, would it cause any performance issues ?

**Reply by gorhill** [reply]
Not an issue when `:watch-attr()` applies to the actual element(s) being targeted. Attribute changes outside these elements won't trigger the mutation observer.

**Comment by Yuki2718** [comment]
On `welovemanga.one` with Chrome, `welovemanga.*##body:remove-attr(onclick)` does not work - onclick is removed but popup still occurs. ~~`welovemanga.*##+js(ra, onclick, body)` works fine~~.

**Reply by uBlock-user** [reply]
`welovemanga.one##+js(nowoif)` ?

**Reply by Yuki2718** [reply]
https://github.com/uBlockOrigin/uAssets/commit/e4a3e896cf6914c85ec5ae85995a33ce5f989f89

**Reply by gorhill** [reply]
Odd, both filters do the same, they call `removeAttribute('onclick')`, and I can see with either that the `onclick` attribute is gone.

However, with `:remove-attr()`, the `body.onclick` event listener is not removed. `removeAttribute()` from the dev console didn't work either, but I could finally remove the listener using `setAttribute('onclick', '')`. All work as expected in Firefox, so this seems to be a Chromium quirk.

What I could do to address this Chromium quirk, is to call `setAttribute(..., '')` just before calling `removeAttribute()` -- hopefully without bad side effects.

---

Adding `setAttribute(..., '')` does not work either. So I think I understand the behavior is because with `ra.js`, the attribute removal is made from the page world, while with `remove-attr`, the attribute removal is made from the content script world.

**Reply by Yuki2718** [reply]
Even with `##+js(ra, onclick, body)` I occasionally got popup, but less degree as `:remove-attr()` 100% failed.

**Reply by gorhill** [reply]
Given there is the `nowoif` solution, I will still consider `ra` to be deprecated in the long term, and `nowoif` is also preferable anyway because it's properly imported into uBOL, while `ra` is not (`remove-attr` would be also properly imported).

**Reply by ghost** [reply]
>the attribute removal is made from the page world, while with remove-attr, the attribute removal is made from the content script world.

Wondering, can this be the cause of why YT or Twitch video still plays after `video` element is removed from DOM by `:remove()`? For example:

    ! https://www.youtube.com/user/LinusTechTips
    youtube.com##ytd-channel-video-player-renderer:remove()


I was never able to figure out what is the source of the audio which still plays.

***

Manually removing this element in Dev Tools works like expected breaking the playback, uBO filter does not.

***

https://www.reddit.com/r/uBlockOrigin/comments/106p2ho/how_do_you_block_this_crap_on_youtube/

***

The whole YT interface is build out of custom elements like `<ytd-channel-video-player-renderer>`, so maybe it match what https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#content_script_environment says.

**Reply by joridiculous** [reply]
This seems to work: https://www.reddit.com/r/uBlockOrigin/comments/goopft/youtube_channel_trailers/frh43am/
(In one of the google search links you posted in my reddit thread)

Mind im on an older Chrome, and some developer version 1.45.3b12

**Reply by uBlock-user** [reply]
> with `ra.js`, the attribute removal is made from the page world, while with `remove-attr`, the attribute removal is made from the content script world.

So is this why `:remove()` doesn't work on inline javascripts ?

**Reply by krystian3w** [reply]
Maybe due race condition if are executed with same speed like meta refresh?

**Comment by mapx-** [comment]
anti right click:

working => `warning.or.kr##+js(ra, oncontextmenu)`

not working => `warning.or.kr##body:remove-attr(oncontextmenu)`
(even if it seems `oncontextmenu` is removed)

tested in chrome
@gorhill

**Reply by uBlock-user** [reply]
Due to https://github.com/uBlockOrigin/uBlock-issues/discussions/2403#discussioncomment-4625517

>  So I think I understand the behavior is because with ra.js, the attribute removal is made from the page world, while with remove-attr, the attribute removal is made from the content script world.
